README.md
# <img alt='' src='https://gitea.com/repo-avatars/550-80a3a8c2ab0e2c2d69f296b7f8582485' height="40"/> *T E A*

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT) [![Release](https://raster.shields.io/badge/dynamic/json.svg?label=release&url=https://gitea.com/api/v1/repos/gitea/tea/releases&query=$[0].tag_name)](https://gitea.com/gitea/tea/releases) [![Build Status](https://drone.gitea.com/api/badges/gitea/tea/status.svg)](https://drone.gitea.com/gitea/tea) [![Join the chat at https://img.shields.io/discord/322538954119184384.svg](https://img.shields.io/discord/322538954119184384.svg)](https://discord.gg/Gitea) [![Go Report Card](https://goreportcard.com/badge/code.gitea.io/tea)](https://goreportcard.com/report/code.gitea.io/tea) [![GoDoc](https://godoc.org/code.gitea.io/tea?status.svg)](https://godoc.org/code.gitea.io/tea)

### The official CLI for Gitea

![demo gif](./demo.gif)

```
   tea - command line tool to interact with Gitea
   version 0.8.0-preview

 USAGE
   tea command [subcommand] [command options] [arguments...]

 DESCRIPTION
   tea is a productivity helper for Gitea. It can be used to manage most entities on
   one or multiple Gitea instances & provides local helpers like 'tea pr checkout'.
   
   tea tries to make use of context provided by the repository in $PWD if available.
   tea works best in a upstream/fork workflow, when the local main branch tracks the
   upstream repo. tea assumes that local git state is published on the remote before
   doing operations with tea.    Configuration is persisted in $XDG_CONFIG_HOME/tea.

 COMMANDS
   help, h  Shows a list of commands or help for one command
   ENTITIES:
     issues, issue, i                  List, create and update issues
     pulls, pull, pr                   Manage and checkout pull requests
     labels, label                     Manage issue labels
     milestones, milestone, ms         List and create milestones
     releases, release, r              Manage releases
     times, time, t                    Operate on tracked times of a repository's issues & pulls
     organizations, organization, org  List, create, delete organizations
     repos, repo                       Show repository details
     comment, c                        Add a comment to an issue / pr
   HELPERS:
     open, o                         Open something of the repository in web browser
     notifications, notification, n  Show notifications
     clone, C                        Clone a repository locally
   SETUP:
     logins, login                  Log in to a Gitea server
     logout                         Log out from a Gitea server
     shellcompletion, autocomplete  Install shell completion for tea
     whoami                         Show current logged in user

 OPTIONS
   --help, -h     show help (default: false)
   --version, -v  print the version (default: false)

 EXAMPLES
   tea login add                       # add a login once to get started

   tea pulls                           # list open pulls for the repo in $PWD
   tea pulls --repo $HOME/foo          # list open pulls for the repo in $HOME/foo
   tea pulls --remote upstream         # list open pulls for the repo pointed at by
                                       # your local "upstream" git remote
   # list open pulls for any gitea repo at the given login instance
   tea pulls --repo gitea/tea --login gitea.com

   tea milestone issues 0.7.0          # view open issues for milestone '0.7.0'
   tea issue 189                       # view contents of issue 189
   tea open 189                        # open web ui for issue 189
   tea open milestones                 # open web ui for milestones

   # send gitea desktop notifications every 5 minutes (bash + libnotify)
   while :; do tea notifications --mine -o simple | xargs -i notify-send {}; sleep 300; done

 ABOUT
   Written & maintained by The Gitea Authors.
   If you find a bug or want to contribute, we'll welcome you at https://gitea.com/gitea/tea.
   More info about Gitea itself on https://gitea.io.
```

- [Compare features with other git forge CLIs](./FEATURE-COMPARISON.md)
- tea uses [code.gitea.io/sdk](https://code.gitea.io/sdk) and interacts with the Gitea API.

## Installation

There are different ways to get `tea`:

1. Install via your system package manager:
    - macOS via `brew` (gitea-maintained):
      ```sh
      brew tap gitea/tap https://gitea.com/gitea/homebrew-gitea
      brew install tea
      ```
    - arch linux ([gitea-tea-git](https://aur.archlinux.org/packages/gitea-tea-git), thirdparty)
    - alpine linux ([tea](https://pkgs.alpinelinux.org/packages?name=tea&branch=edge), thirdparty)

2. Use the prebuilt binaries from [dl.gitea.io](https://dl.gitea.io/tea/)

3. Install from source: [see *Compilation*](#compilation)

4. Docker (thirdparty): [tgerczei/tea](https://hub.docker.com/r/tgerczei/tea)

## Compilation

Make sure you have a current go version installed (1.13 or newer).

- To compile the source yourself with the recommended flags & tags:
  ```sh
  git clone https://gitea.com/gitea/tea.git # or: tea clone gitea.com/gitea/tea ;)
  cd tea
  make
  ```
  Note that GNU Make (gmake on OpenBSD) is required.

- For a quick installation without `git` & `make`, set $version and exec:
  ```sh
  go install code.gitea.io/tea@${version}
  ```

## Contributing

Fork -> Patch -> Push -> Pull Request

- `make test` run testsuite
- `make vet`  run checks (check the order of imports; preventing failure on CI pipeline beforehand)
- ... (for other development tasks, check the `Makefile`)

**Please** read the [CONTRIBUTING](CONTRIBUTING.md) documentation, it will tell you about internal structures and concepts.

## License

This project is under the MIT License. See the [LICENSE](LICENSE) file for the
full license text.


CHANGELOG.md
# Changelog

## [v0.9.1](https://gitea.com/gitea/tea/releases/tag/v0.9.1) - 2023-02-15

* BUGFIXES
  * Print pull dont crash if it has TeamReviewRequests (#517)

## [v0.9.0](https://gitea.com/gitea/tea/releases/tag/v0.9.0) - 2022-09-13

* BREAKING
  * Rename master branch to main (#495)
  * Return RFC3339 UTC timestamps for machine-readable output (#470)
* FEATURES
  * Allow editing multiline prompts with external text editor (#429)
  * Add `tea admin user list` (#427)
  * Add `tea whoami` command (#426)
  * Add `tea org create <name>` (#420)
  * Add `tea clone` (#411)
  * Add `tea repo fork` (#410)
  * Add `tea repo create-from-template` (#408)
* BUGFIXES
  * Fetch all items where needed. (#475)
  * Fix running in repos without remote (#472)
  * Add TSV to machine-readable formats (#467)
  * Fix create milestone with deadline bug (#462)
  * Fix resolving of URLs in markdown (#401)
* ENHANCEMENTS
  * Don't emit ANSI sequences when not emitting to TTY for markdown (#491)
  * Show more version info (#486)
  * Add preference `flag_defaults.remote`, refactor (#466)
  * Add `--fields` to notification & milestone listings (#422)
  * PR listing: add --fields & expose additional fields (#415)
  * Add more flags to `tea repo create` (#409)
  * Implement more issue filters (#400)
* MISC
  * Simplify build & update installation instructions (#437)
  * Clarify command descriptions when no arguments are taken (#496)
  * Improve Documentation (#433)
  * Use golang v1.18 and drop vendor folder (#478)
  * Correct spelling of "wether" to "whether" in usage output (#453)

## [v0.8.0](https://gitea.com/gitea/tea/releases/tag/v0.8.0) - 2021-09-22

* BREAKING
  * `tea notifications --all` has moved to `tea notifications --mine` (#389)
  * `tea notifications` now only works with the context of a remote repo. (#389)
    To run this outside of a local git dir, run either tea n `--mine` or `tea n --repo <my/repo>`
* FEATURES
  * Add `tea pr merge` (#348)
* BUGFIXES
  * Don't skip reading the local repo when `--repo` specifies a repo slug (#398)
  * Fix adding login without token on private instances (#392)
  * Correctly match login by ssh host with port (#391)
  * Fix printing issue deadline (#388)
  * Return useful error on wrong sshkey path (#374)
  * Fix parsing of `--description` for issue/pr create (#371)
  * Add missing flags (#369)
  * Check negative limit command parameter (#358) (#359)
  * Add missing flags to org & labels subcommands (#357)
* ENHANCEMENTS
  * Don't require a body for comment PR reviews (#399)
  * Accept more main branch names for login detection (#396)
  * Make local repo optional for `tea pr create`(#393)
  * Notifications Add State Field (#384)
  * Improve error messages (#370)
  * Add tab completion for fish shell (#364)
  * Text editor selection: follow unix defacto standards (#356)
* MISC
  * Update Dependencies (#390)

## [v0.7.1](https://gitea.com/gitea/tea/releases/tag/v0.7.1) - 2021-08-27

* BUILD
  * Enable release builds for darwin/arm64 (#360)

## [v0.7.0](https://gitea.com/gitea/tea/releases/tag/v0.7.0) - 2021-03-12

* BREAKING
  * `tea issue create`: move `-b` flag to `-d` (#331)
  * Drop `tea notif` shorthand in favor of `tea n` (#307)
* FEATURES
  * Add commands for reviews (#315)
  * Add `tea comment` and show comments of issues/pulls (#313)
  * Add interactive mode for `tea milestone create` (#310)
  * Add command to install shell completion (#309)
  * Implement PR closing and reopening (#304)
  * Add interactive mode for `tea issue create` (#302)
* BUGFIXES
  * Introduce workaround for missing pull head sha (#340)
  * Don't exit if we can't find a local repo with a remote matching to a login (#336)
  * Don't push before creating a pull (#334)
  * InitCommand() robustness (#327)
  * `tea comment`: handle piped stdin (#322)
* ENHANCEMENTS
  * Allow checking out PRs with deleted head branch (#341)
  * Markdown renderer: detect terminal width, resolve relative URLs (#332)
  * Add more issue / pr creation parameters (#331)
  * Improve `tea time` (#319)
  * `tea pr checkout`: dont create local branches (#314)
  * Add `tea issues --fields`, allow printing labels (#312)
  * Add more command shorthands (#307)
  * Show PR CI status (#306)
  * Make PR workflow helpers more robust (#300)

## [v0.6.0](https://gitea.com/gitea/tea/releases/tag/v0.6.0) - 2020-12-11

* BREAKING
  * Add `tea repos search`, improve repo listing (#215)
  * Add Detail View for Login (#212)
* FEATURES
  * Add interactive mode for `tea pr create` (#279)
  * Add organization delete command (#270)
  * Add organization list command (#264)
* BUGFIXES
  * Forces needed arguments to `tea ms issues` (#297)
  * Subcommands work outside of git repos (#285)
  * Fix repo flag ignores local repo for login detection (#285)
  * Improve ssh handling (#277)
  * Issue create return web url (#257)
  * Support prerelease gitea instances (#252)
  * Fix `tea pr create` within same repo (#248)
  * Handle login name case-insensitive on all comands (#227)
* ENHANCEMENTS
  * Add `tea login delete` (#296)
  * Release delete: add --delete-tag & --confirm (#286)
  * Sorted milestones list (#281)
  * Pull clean & checkout use token for http(s) auth (#275)
  * Show more infos in pull detail view (#271)
  * Specify fields to print on `tea repos list` (#223)
  * Print times in local timezone (#217)
  * Issue create/edit print details (#214)
  * Improve `tea logout` (#213)
  * Added a shorthand for notifications (#209)
  * Common subcommand naming scheme (#208)
  * `tea pr checkout`: fetch via ssh if available (#192)
  * Major refactor of codebase
* BUILD
  * Use gox to cross-compile (#274)
* DOCS
  * Update Docs to new code structure (#247)

## [v0.5.0](https://gitea.com/gitea/tea/releases/tag/v0.5.0) - 2020-09-27

* BREAKING
  * Add Login Manage Functions (#182)
* FEATURES
  * Add Release Subcomands (#195)
  * Render Markdown and colorize labels table (#181)
  * Add BasicAuth & Interactive for Login (#174)

main.go
// Copyright 2020 The Gitea Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

// Tea is command line tool for Gitea.
package main // import "code.gitea.io/tea"

import (
	"fmt"
	"os"
	"runtime"
	"strings"

	"code.gitea.io/tea/cmd"

	"github.com/urfave/cli/v2"
)

// Version holds the current tea version
var Version = "development"

// Tags holds the build tags used
var Tags = ""

// SDK holds the sdk version from go.mod
var SDK = ""

func main() {
	// make parsing tea --version easier, by printing /just/ the version string
	cli.VersionPrinter = func(c *cli.Context) { fmt.Fprintln(c.App.Writer, c.App.Version) }

	app := cli.NewApp()
	app.Name = "tea"
	app.Usage = "command line tool to interact with Gitea"
	app.Description = appDescription
	app.CustomAppHelpTemplate = helpTemplate
	app.Version = formatVersion()
	app.Commands = []*cli.Command{
		&cmd.CmdLogin,
		&cmd.CmdLogout,
		&cmd.CmdAutocomplete,
		&cmd.CmdWhoami,

		&cmd.CmdIssues,
		&cmd.CmdPulls,
		&cmd.CmdLabels,
		&cmd.CmdMilestones,
		&cmd.CmdReleases,
		&cmd.CmdTrackedTimes,
		&cmd.CmdOrgs,
		&cmd.CmdRepos,
		&cmd.CmdAddComment,

		&cmd.CmdOpen,
		&cmd.CmdNotifications,
		&cmd.CmdRepoClone,

		&cmd.CmdAdmin,
	}
	app.EnableBashCompletion = true
	err := app.Run(os.Args)
	if err != nil {
		// app.Run already exits for errors implementing ErrorCoder,
		// so we only handle generic errors with code 1 here.
		fmt.Fprintf(app.ErrWriter, "Error: %v\n", err)
		os.Exit(1)
	}
}

func formatVersion() string {
	version := fmt.Sprintf("Version: %s\tgolang: %s",
		bold(Version),
		strings.ReplaceAll(runtime.Version(), "go", ""))

	if len(Tags) != 0 {
		version += fmt.Sprintf("\tbuilt with: %s", strings.Replace(Tags, " ", ", ", -1))
	}

	if len(SDK) != 0 {
		version += fmt.Sprintf("\tgo-sdk: %s", SDK)
	}

	return version
}

var appDescription = `tea is a productivity helper for Gitea. It can be used to manage most entities on
one or multiple Gitea instances & provides local helpers like 'tea pr checkout'.

tea tries to make use of context provided by the repository in $PWD if available.
tea works best in a upstream/fork workflow, when the local main branch tracks the
upstream repo. tea assumes that local git state is published on the remote before
doing operations with tea.    Configuration is persisted in $XDG_CONFIG_HOME/tea.
`

var helpTemplate = bold(`
   {{.Name}}{{if .Usage}} - {{.Usage}}{{end}}`) + `
   {{if .Version}}{{if not .HideVersion}}version {{.Version}}{{end}}{{end}}

 USAGE
   {{if .UsageText}}{{.UsageText}}{{else}}{{.HelpName}}{{if .Commands}} command [subcommand] [command options]{{end}} {{if .ArgsUsage}}{{.ArgsUsage}}{{else}}[arguments...]{{end}}{{end}}{{if .Description}}

 DESCRIPTION
   {{.Description | nindent 3 | trim}}{{end}}{{if .VisibleCommands}}

 COMMANDS{{range .VisibleCategories}}{{if .Name}}
   {{.Name}}:{{range .VisibleCommands}}
     {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{else}}{{range .VisibleCommands}}
   {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{end}}{{end}}{{end}}{{if .VisibleFlags}}

 OPTIONS
   {{range $index, $option := .VisibleFlags}}{{if $index}}
   {{end}}{{$option}}{{end}}{{end}}

 EXAMPLES
   tea login add                       # add a login once to get started

   tea pulls                           # list open pulls for the repo in $PWD
   tea pulls --repo $HOME/foo          # list open pulls for the repo in $HOME/foo
   tea pulls --remote upstream         # list open pulls for the repo pointed at by
                                       # your local "upstream" git remote
   # list open pulls for any gitea repo at the given login instance
   tea pulls --repo gitea/tea --login gitea.com

   tea milestone issues 0.7.0          # view open issues for milestone '0.7.0'
   tea issue 189                       # view contents of issue 189
   tea open 189                        # open web ui for issue 189
   tea open milestones                 # open web ui for milestones

   # send gitea desktop notifications every 5 minutes (bash + libnotify)
   while :; do tea notifications --mine -o simple | xargs -i notify-send {}; sleep 300; done

 ABOUT
   Written & maintained by The Gitea Authors.
   If you find a bug or want to contribute, we'll welcome you at https://gitea.com/gitea/tea.
   More info about Gitea itself on https://gitea.io.
`

func bold(t string) string {
	return fmt.Sprintf("\033[1m%s\033[0m", t)
}
